helix-vcs/src/diff.rs

```diff
@@ -0,0 +1,218 @@
+        } else {
+            self.update_document_impl(doc)
+        }
+    }
+
+    pub fn update_diff_base(&self, diff_base: Rope) -> bool {
+        if self.inverted {
+            self.update_document_impl(diff_base)
+        } else {
+            self.update_diff_base_impl(diff_base)
+        }
+    }
+
+    pub fn update_document_impl(&self, doc: Rope) -> bool {
+        self.channel.send(Event::UpdateDocument(doc)).is_ok()
+    }
+
+    pub fn update_diff_base_impl(&self, diff_base: Rope) -> bool {
+        self.channel.send(Event::UpdateDiffBase(diff_base)).is_ok()
+    }
+}
+
+// TODO configuration
+const DIFF_DEBOUNCE_TIME: u64 = 100;
+const ALGORITHM: Algorithm = Algorithm::Histogram;
+const MAX_DIFF_LINES: usize = u16::MAX as usize;
+// cap average line length to 128 for files with MAX_DIFF_LINES
+const MAX_DIFF_BYTES: usize = MAX_DIFF_LINES * 128;
+
+/// A single change in a file potentially sppaning multiple lines
+/// Hunks produced by the differs are always ordered by their position
+/// in the file and non-overlapping.
+/// Specifically for any two hunks `x` and `y` the following properties hold:
+///
+/// ``` no_compile
+/// assert!(x.before.end <= y.before.start);
+/// assert!(x.after.end <= y.after.start);
+/// ```
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct Hunk {
+    pub before: Range<u32>,
+    pub after: Range<u32>,
+}
+
+impl Hunk {
+    /// Can be used instead of `Option::None` for better performance
+    /// because lines larger then `i32::MAX` are not supported by imara-diff anways.
+    /// Has some nice properties where it usually is not necessary to check for `None` seperatly:
+    /// Empty ranges fail contains checks and also faills smaller then checks.
+    pub const NONE: Hunk = Hunk {
+        before: u32::MAX..u32::MAX,
+        after: u32::MAX..u32::MAX,
+    };
+
+    /// Inverts a change so that `before`
+    pub fn invert(&self) -> Hunk {
+        Hunk {
+            before: self.after.clone(),
+            after: self.before.clone(),
+        }
+    }
+
+    pub fn is_pure_insertion(&self) -> bool {
+        self.before.is_empty()
+    }
+
+    pub fn is_pure_removal(&self) -> bool {
+        self.after.is_empty()
+    }
+}
+
+/// A list of changes in a file sorted in ascending
+/// non-overlapping order
+#[derive(Debug)]
+pub struct FileHunks<'a> {
+    hunks: MutexGuard<'a, Vec<Hunk>>,
+    inverted: bool,
+}
+
+impl FileHunks<'_> {
+    pub fn is_inverted(&self) -> bool {
+        self.inverted
+    }
+
+    /// Returns the `Hunk` for the `n`th change in this file.
+    /// if there is no `n`th change  `Hunk::NONE` is returned instead.
+    pub fn nth_hunk(&self, n: u32) -> Hunk {
+        match self.hunks.get(n as usize) {
+            Some(hunk) if self.inverted => hunk.invert(),
+            Some(hunk) => hunk.clone(),
+            None => Hunk::NONE,
+        }
+    }
+
+    fn next_hunk_impl<const INVERT: bool, const REVERSE: bool>(&self, line: u32) -> Option<u32> {
```

Comment:
I think this would be clearer if it used a `helix_core::movement::Direction` parameter than a const generic for `REVERSE`. Similarly, can't we use `Self::is_inverted` or `self.inverted` directly instead of `INVERT`?

IIUC, the const generic is faster because the compiler knows the value at compile-time and can use that knowledge for optimization but I'm not sure this is a worthy tradeoff unless the performance improvement is noticeable

Reply:
It's mostly a style choice for `REVERSE`. If you look at the actual implementation of `next_hunk_impl`, the `REVERSE` and non `REVERSE` case are basically two separate functions with a shared structure. I thought the details would be easier to grasp if they are merged but in my mind these are very much separate functions that I wouldn't merge with a bool conditions. I can imagine a couple usecases in the future that might be a bit more performance sensitive that might only want to call `next_hunk` or `prev_hunk`.

I guess I view the fact that `next_hunk` and `prev_hunk` are internally the same function mostly as an implementation detail and something I would want to expose by making it part of the API.

Once I was already using const generics I guess it didn't hurt to add one for `INVERT` aswell.
Both `INVERT` and `REVERSE` are checked as part of the binary search so I also liked the garuntee of the condition being outlined.

If you think it's really ugly I can definitely change this but I actually really like this pattern of using const generics to outline conditions out of functions.

Reply:
I wouldn't mind it so much if it were more common (look at how [Kakoune uses a similar pattern](https://github.com/mawww/kakoune/blob/760a45556cfb8f50813ce8d1712a0a23f86dcf52/src/normal.cc#L2436-L2442) with C++ equivalents) and if we could use enums instead of bools (AFAIK this is planned for the future in the language). But this will be the only use of const generics in the codebase and it's sacrificing what can be a Forward/Backward enum for a bool which we try to push against (https://github.com/helix-editor/helix/pull/1723#discussion_r831987286, https://github.com/helix-editor/helix/pull/4134#discussion_r994061294)

Reply:
I removed the const-generics. For `next`/`prev` I created two seperate functions and an `if` for `inverted`.
I opted for separate functions as I need to match on the direction anyway at the callsite to correctly account for `count` (and as described above), while the implementation are basically two seperate functions anyway